### PR TITLE
Lfc batch

### PR DESCRIPTION
**Pull request by PierreBoyeau**

### CODE REVIEW

File: src/scvi_v2/_model.py [hunk 1]
```diff
@@ -1189,15 +1185,7 @@ def u_statistic(rank_x):
...
-            if store_lfc:
-                set_shape = (mc_samples, n_covariates_require_lfc)
-                stacked_rngs_de = [
-                    self._generate_stacked_rngs(n_sets=set_shape),
-                    self._generate_stacked_rngs(n_sets=set_shape),
-                ]
-            else:
-                stacked_rngs_de = None
-
+            rngs_de = self.module.rngs if store_lfc else None
```
Comment (canergen) [inline src/scvi_v2/_model.py:1077]: Where is this coming from?

File: src/scvi_v2/_model.py [hunk 2]
```diff
@@ -1119,45 +1118,40 @@ def per_sample_inference_fn(pair):
...
-                    return jnp.sum(rank_x, axis=1) - (mc_samples * (mc_samples + 1) / 2)
-
-                u_x = jnp.minimum(u_statistic(rank_x[:, :mc_samples, :, :]), u_statistic(rank_x[:, mc_samples:, :, :]))
-                zscore = (u_x - mc_samples**2 / 2) / jnp.sqrt(mc_samples**2 * (2 * mc_samples + 1) / 12)
-
-                mean_lfcs = lfcs.mean(1)
+                batch_specific_lfc = lfcs.mean(1)
+                lfc_mean = batch_specific_lfc.mean(0)
+                lfc_std = batch_specific_lfc.std(0)
+                pde = (jnp.abs(lfcs) >= delta).mean(1).mean(0)
```
Comment (canergen) [inline src/scvi_v2/_model.py]: What's the interpretation of this value? Is it computationally hard to get something pvalue like as in model.differential_expression?

File: src/scvi_v2/_model.py [hunk 3]
```diff
@@ -1119,45 +1118,40 @@ def per_sample_inference_fn(pair):
...
-                def u_statistic(rank_x):
-                    return jnp.sum(rank_x, axis=1) - (mc_samples * (mc_samples + 1) / 2)
-
-                u_x = jnp.minimum(u_statistic(rank_x[:, :mc_samples, :, :]), u_statistic(rank_x[:, mc_samples:, :, :]))
-                zscore = (u_x - mc_samples**2 / 2) / jnp.sqrt(mc_samples**2 * (2 * mc_samples + 1) / 12)
-
-                mean_lfcs = lfcs.mean(1)
+                batch_specific_lfc = lfcs.mean(1)
+                lfc_mean = batch_specific_lfc.mean(0)
+                lfc_std = batch_specific_lfc.std(0)
```
Comment (canergen) [inline src/scvi_v2/_model.py]: Not sure this is the best way to report variance

File: src/scvi_v2/_model.py (hunk 2 above)
Comment (canergen) [inline src/scvi_v2/_model.py]: Can we convert it to FDR?

File: src/scvi_v2/_model.py [hunk 4]
```diff
@@ -1062,15 +1062,14 @@ def mapped_inference_fn(
             stacked_rngs,
             x,
             sample_index,
-            batch_index,
             continuous_covs,
             cf_sample,
             Amat,
             prefactor,
             n_donors_per_cell,
```
Comment (canergen) [inline src/scvi_v2/_model.py:943]: Let's discuss mean only over admissible. I'm not to clear about it.

File: src/scvi_v2/_model.py (hunk 2 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: I don't think so unfortunately

Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: It's a score that quantifies the probability that the gene is DE

**Comment by PierreBoyeau** [comment]
I have concerns about the last commit about including batch-specific offsets in the design matrix, and the tests don't pass on my side.

File: src/scvi_v2/_model.py [hunk 5]
```diff
@@ -1110,54 +1116,53 @@ def per_sample_inference_fn(pair):
...
-                concat_x = jnp.concatenate([x_0, x_1], axis=1)
-                rank_x = jax.scipy.stats.rankdata(concat_x, method="average", axis=1)
+                betas_offset_ = betas_[:, offset_indices, :, :] if offset_indices else jnp.zeros(
+                    [betas_.shape[0], batch_index_.shape[0], betas_.shape[2], betas_.shape[3]])
 
-                def u_statistic(rank_x):
-                    return jnp.sum(rank_x, axis=1) - (mc_samples * (mc_samples + 1) / 2)
+                f_ = jax.vmap(h_inference_fn, in_axes=(0, None, 0), out_axes=0)  # fn over MC samples
+                f_ = jax.vmap(f_, in_axes=(1, None, None), out_axes=1)  # fn over covariates
+                f_ = jax.vmap(f_, in_axes=(None, 0, 1), out_axes=0)  # fn over batches
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py:1028]: how are we sure that batch_indices and batch_offset_eps batch channels are in the same order?

File: src/scvi_v2/_model.py [hunk 6]
```diff
@@ -1331,9 +1326,17 @@ def _construct_design_matrix(
         if normalize_design_matrix:
             Xmat = (Xmat - Xmat.min(axis=0)) / (1e-6 + Xmat.max(axis=0) - Xmat.min(axis=0))
         if offset_design_matrix:
-            Xmat = np.concatenate([np.ones((Xmat.shape[0], 1)), Xmat], axis=1)
-            Xmat_names = np.concatenate([np.array(["offset"]), Xmat_names])
-            Xmat_dim_to_key = np.concatenate([np.array(["offset"]), Xmat_dim_to_key])
+            cov = self.donor_info['_scvi_batch']
```
Comment (PierreBoyeau) [inline src/scvi_v2/_model.py]: When there are only two batches in the data, this only adds one column to the design matrix. Maybe this is why the tests don't pass?

File: src/scvi_v2/_model.py (hunk 3 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: What would you suggest?

File: src/scvi_v2/_model.py (hunk 5 above)
Comment (canergen) [reply src/scvi_v2/_model.py:1028]: Oh, that's a good catch. One_hot_encoding doesn't keep the order when we create the columns. We should fix this.

File: src/scvi_v2/_model.py [hunk 7]
```diff
@@ -1098,9 +1100,14 @@ def per_sample_inference_fn(pair):
                     return inference_fn(rngs, cf_sample)
 
                 # eps_ has shape (mc_samples, n_cells, n_donors, n_latent)
-                eps_ = jax.lax.transpose(jax.lax.map(per_sample_inference_fn, (stacked_rngs, cf_sample)), (1, 0, 2))
-            eps = (eps_ - eps_.mean(axis=2, keepdims=True)) / (1e-6 + eps_.std(axis=2, keepdims=True))  # over donors
+                eps_ = jax.lax.transpose(jax.lax.map(per_sample_inference_fn, (stacked_rngs, cf_sample)), (1, 2, 0, 3))
+            # admissible_donors_mat_ = jnp.expand_dims(admissible_donors_mat, [0, 3])
+            batch_weights = jnp.einsum("nd,da->na", admissible_donors_mat, Xmat[:, offset_indices]).mean(0)
```
Comment (canergen) [inline src/scvi_v2/_model.py]: This one should fail if offset_indices is None.

Comment (PierreBoyeau) [reply src/scvi_v2/_model.py]: TODO: Fix

File: src/scvi_v2/_module.py [hunk 8]
```diff
@@ -490,9 +490,8 @@ def generative(self, z, library, batch_index, label_index, continuous_covs):
         h = px.mean / library_exp
 
         if self.u_prior_mixture:
-            one_hot_labels = jax.nn.one_hot(label_index, self.n_labels)
-            # kind of arbitrary factor, I found it well balanced but might require adjustment.
-            cats = dist.Categorical(logits=10 * one_hot_labels + self.u_prior_logits)
+            offset = 10.0 * jax.nn.one_hot(label_index, self.n_labels) if self.n_labels >= 2 else 0.0
```
Comment (justjhong) [inline src/scvi_v2/_module.py:493]: what is this?

File: src/scvi_v2/_model.py (hunk 7 above)
Comment (justjhong) [reply src/scvi_v2/_model.py]: do we need a fix here for when admissible cells is zero for a batch? Or does the contribution go to zero as expected

File: src/scvi_v2/_module.py (hunk 8 above)
Comment (canergen) [reply src/scvi_v2/_module.py:493]: The cell-type bias used for IBD and Immuneaging. It associates one cell-type with one of the MoG modes.

File: src/scvi_v2/_model.py (hunk 7 above)
Comment (canergen) [reply src/scvi_v2/_model.py]: I'm pretty sure it goes to zero as expected.

**Pull request merged**